### PR TITLE
Show sparkline evolution figures for visits, goals and ecommerce overviews

### LINKED ISSUE #10716: 3.0b1: show evolution as number on all overview sites
Would be handy to see evolution as number (xxx%) on all overview sites
(like it's already on referres)

2016-10-07_18h03_40

### PR DESCRIPTION
**Pull request by bx80**
### Description:

Fixes #10716

This PR adds evolution percentages and up/down icons to the end of the sparklines shown on the overview screens for visits, goals and ecommerce.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by bx80** [comment]
Work in progress - visits and goals overviews are done, ecommerce renders sparklines a little differently and is not implemented yet.

**Comment by bx80** [comment]
The Ecommerce overview spark lines have been implemented using a custom template rather than a visualization, probably to allow special handling for the abandoned cart goal metrics. There isn't enough time allocated for this issue to rework the overview to use a visualization, so instead I've chosen a pragmatic approach and rendered the evolution values in the custom spark line template. This isn't ideal but didn't take much time and will provide evolution metrics in one of the more useful places until things can be reworked at some future time.

In order for the override method `Ecommerce.getMetricsForGoal()` to gain access to the data row used in the parent `Goals.getMetricsForGoal()` I've added an optional parameter to the parent method to allow the preloaded data row to be passed by the child method. I thought this preferable to replicating the entire parent method, but open to better ideas! :slightly_smiling_face:

File: plugins/Goals/Reports/Get.php
Comment (sgiehl) [inline plugins/Goals/Reports/Get.php]: Shouldn't that one be a formatted date including the period? Currently when viewing a week or month, the tooltip will always show the first day of the compared periods each, which isn't correct.

File: plugins/VisitsSummary/Reports/Get.php
Comment (sgiehl) [inline plugins/VisitsSummary/Reports/Get.php]: Same for this tooltip: date should be formatted with period.

File: plugins/VisitsSummary/Reports/Get.php
Comment (sgiehl) [inline plugins/VisitsSummary/Reports/Get.php]: I guess to get this and other metrics to work we would need to request the data with `format_metrics=0` and format the metrics manually before they are displayed. This might need a bit more effort but would also fix some other issues that might currently occur when formatted values are compared.

File: plugins/CoreVisualizations/Visualizations/Sparklines.php
Comment (sgiehl) [inline plugins/CoreVisualizations/Visualizations/Sparklines.php]: Might be good to improve/update the documentation of that callback method here:

https://github.com/matomo-org/matomo/blob/c11469e4199405628e97650914f0cdbdea94dac7/plugins/CoreVisualizations/Visualizations/Sparklines/Config.php#L49-L56

File: plugins/VisitsSummary/Reports/Get.php
Comment (sgiehl) [inline plugins/VisitsSummary/Reports/Get.php]: We maybe should use the translation contained in `Metrics::getDefaultMetricTranslations()` if available as default. Currently the sparkline added by the bandwidth plugin falls back to use `Visits` in the tooltip, where it actually should show `transfered bytes`

**Review by sgiehl — CHANGES_REQUESTED** [review]
Had a rough look through the changes. The approach looks fine, but we might have a similar issue like we had a plenty of times before. We always request the data formatted from the API and afterwards try to compare the data. That often fails for languages that use other formattings than english. I guess at some point we need to decide to put a lot more effort into that topic and start changing that behavior to request unformatted data and format it when it is shown. Not sure if we should do that now. But I'm pretty sure that for some languages this new evolutions might show incorrect values....

File: plugins/Goals/Reports/Get.php
Comment (bx80) [reply plugins/Goals/Reports/Get.php]: I've adjusted the sparkline evolution tooltip code to build period objects and use them to generate a  properly formatted period date.

File: plugins/CoreVisualizations/Visualizations/Sparklines.php
Comment (bx80) [reply plugins/CoreVisualizations/Visualizations/Sparklines.php]: :+1: Done

File: plugins/VisitsSummary/Reports/Get.php
Comment (bx80) [reply plugins/VisitsSummary/Reports/Get.php]: Done - that's a much nicer approach :slightly_smiling_face:

File: plugins/VisitsSummary/Reports/Get.php
Comment (bx80) [reply plugins/VisitsSummary/Reports/Get.php]: I've committed a change which makes sparkline API calls with the `format_metrics=0` parameter, this seems to be working well and the `compute_evolution()` callback now received unformatted data allowing time, percent and money metrics to be safely calculated.

The main sparkline visualization `CoreVisualizations\Visualizations\Sparklines` has been updated to format the sparkline values using the appropriate Metric class retrieved from the report.

The goals and visits summary overview reports have been updated to format the sparkline evolution tooltips by finding the Metric object for the column and using it's format method.

Ecommerce doesn't use a report or visualization and as noted before it needs some rework. For now I've done a quick fix to manually format the metrics by name.

The bandwidth plugin was pushing values into the datatable pre-formatted, so I've committed a change there to provide unformatted values instead and also added a new Metric class for formatting the overall bandwidth metric. It seems to be working well on the visit summary overview.

File: core/ViewDataTable/Request.php
Comment (sgiehl) [inline core/ViewDataTable/Request.php]: Looking at the usages of that method it seems to make sense to let `$extraParams` overwrite `$requestArray`.
Would you mind adding that param to the doc block and mention that it will be used to overwrite the request params.
Maybe we could even rename the parameter to something like `$forcedParams`, so it's directly clear what it does.

File: plugins/CoreVisualizations/Visualizations/Sparklines.php
Comment (sgiehl) [inline plugins/CoreVisualizations/Visualizations/Sparklines.php]: Guess we could also change the base method instead - if that doesn't break anything. So we don't need to overwrite it here. There might be use cases later where forcing certain parameters might be useful.

File: plugins/Ecommerce/Controller.php
Comment (sgiehl) [inline plugins/Ecommerce/Controller.php]: Not that important, but for comparing strings we actually can always use `===`

File: plugins/CoreVisualizations/Visualizations/Sparklines/Config.php
Comment (sgiehl) [inline plugins/CoreVisualizations/Visualizations/Sparklines/Config.php]: Looking through the code it actually looks like that this second parameter is never really used. Wondering if we shouldn't simply remove it...

File: plugins/CoreVisualizations/Visualizations/Sparklines/Config.php
Comment (sgiehl) [inline plugins/CoreVisualizations/Visualizations/Sparklines/Config.php:64]: In this file we need to change this lines:
https://github.com/matomo-org/matomo/blob/a23094130dcf195978d7404ba6cc385e2094aecd/plugins/CoreVisualizations/Visualizations/Sparklines/Config.php#L305-L308
and add something like
```
'trend' => $evolution['currentValue'] - $evolution['pastValue'],
```

Otherwise the trend formatting in the sparklines template won't work in other languages. In Persian (fa) for example this currently looks like this:

image

File: plugins/Ecommerce/templates/getSparklines.twig
Comment (sgiehl) [inline plugins/Ecommerce/templates/getSparklines.twig]: a non-negative percentage has a prepended `+` for other sparklines. If it isn't possible to use the sparkline marco here for some reason we should at least have the same behavior here.

File: plugins/Ecommerce/Controller.php
Comment (sgiehl) [inline plugins/Ecommerce/Controller.php:118]: I know that you did it that way as the ecommerce sparklines did have a special handling in the code already.
Are you able to say if there is (still) a good reason for having a special handling there?
Can't we use the sparkline handling that is used everywhere else or at least the twig macros if the layout differs?
It for sure would be a bit more effort to clean & improve the code. But personally I would prefer doing that now instead of maybe creating an issue to do it later. At some point we need to start refactoring such older code otherwise our code quality won't get better any time.
Also it might be easier to do that now, as you are already into that code. At a later point someone always would need to think that all through again.
@justinvelluppillai What do you think on that?

**Review by sgiehl — CHANGES_REQUESTED** [review]
Already goes in the right direction, but there are still a couple of issues with the latest changes.

File: core/ViewDataTable/Request.php
Comment (bx80) [reply core/ViewDataTable/Request.php]: Done :+1:

File: plugins/CoreVisualizations/Visualizations/Sparklines.php
Comment (bx80) [reply plugins/CoreVisualizations/Visualizations/Sparklines.php]: I've updated the base method, it's used by the Visualization class too which hopefully won't be a problem. Also added a commit to the MediaAnalytics plugin adjust the inherited class signature.

File: plugins/CoreVisualizations/Visualizations/Sparklines/Config.php
Comment (bx80) [reply plugins/CoreVisualizations/Visualizations/Sparklines/Config.php]: It was added for this change initially, but with the recent updates is no longer needed, I've removed it and updated the callback function documentation.

File: plugins/CoreVisualizations/Visualizations/Sparklines/Config.php
Comment (bx80) [reply plugins/CoreVisualizations/Visualizations/Sparklines/Config.php:64]: Done, I tested it with a فارسی locale and the trend indicators are now shown correctly :+1:

File: plugins/Ecommerce/templates/getSparklines.twig
Comment (bx80) [reply plugins/Ecommerce/templates/getSparklines.twig]: I've updated the template to show a `+` for positive percentages, it should now be consistent with the standard sparkline macro.

File: plugins/Ecommerce/Controller.php
Comment (bx80) [reply plugins/Ecommerce/Controller.php:118]: It would be great to get the sparklines standardized everywhere and get rid of the special handling, it will likely cause other issues in the future :+1:

At the start of this change I did spend a bit of time evaluating how much work would be required to convert ecommerce to use standard sparkline visualizations and reports. The layout for ecommerce sparklines is different with three columns instead of two (which don't flow very well) and additional icons. Another issue is that the ecommerce metrics are repurposed goal metrics so they have their labels and descriptions customised. This is likely why special handling was used.

My take was that either the standard sparkline visualizations and macros would have to be enhanced to support more layout configuration options or the ecommerce overview would need to be redesigned into something more consistent with the rest of the application (and possibly new metric classes added to handle the customisation). Both would take some time.

The original issue set an 8hr time limit on the total change and I'm well beyond that after needing to switch sparkline provided data from formatted to unformatted.

Maybe we could add a separate issue to improve the ecommerce sparklines and prioritize it soon after this change so it can benefit from the time spent here?

File: plugins/CoreVisualizations/Visualizations/Sparklines.php
Comment (bx80) [reply plugins/CoreVisualizations/Visualizations/Sparklines.php]: Actually I've now reverted changes to the `ViewDataTable` and `Visualization` classes as well a the change to MediaAnalytics as it's going to break plugin compatibility, back to overriding the method on the sparkline visualisation.

File: plugins/MediaAnalytics/Visualizations/MediaEvolution.php
Comment (sgiehl) [inline plugins/MediaAnalytics/Visualizations/MediaEvolution.php]: We can't add this file to core.

File: plugins/Ecommerce/Controller.php
Comment (sgiehl) [reply plugins/Ecommerce/Controller.php:118]: I'm aware that it would take some additional time.
Creating additional issues for refactoring might not be my first choice. Based on my experience with other projects such technical dept issues often end up in a big grave of issues, that are almost never worked on. Mostly because as from business point of view they don't have a directly visible business value. (At the point where this would become visible it would be already to late and building something from scratch is often simpler & cheaper)
Nevertheless, if we decide to create new issues for that, we need to find a way to make sure they won't end up like #5985, #13829 or #9312
But maybe lets discuss that in our meeting next week.

File: plugins/Ecommerce/templates/getSparklines.twig
Comment (sgiehl) [inline plugins/Ecommerce/templates/getSparklines.twig]: To be fully correct you need to use `'Intl_NumberSymbolPlus'|translate` instead of a simple `+`, as some languages use a custom plus sign

File: plugins/CoreVisualizations/Visualizations/Sparklines.php
Comment (sgiehl) [reply plugins/CoreVisualizations/Visualizations/Sparklines.php]: Left a comment on the plugin PR. Unless I have something wrong in mind I don't think this would break bc.

**Review by sgiehl — CHANGES_REQUESTED** [review]

File: plugins/MediaAnalytics/Visualizations/MediaEvolution.php
Comment (bx80) [reply plugins/MediaAnalytics/Visualizations/MediaEvolution.php]: Sorry, looks like my submodule wasn't configured correctly. I've reverted this.

File: plugins/Ecommerce/templates/getSparklines.twig
Comment (bx80) [reply plugins/Ecommerce/templates/getSparklines.twig]: Good point, thanks!

**Comment by sgiehl** [comment]
@bx80 I've pushed an update to the bandwidth plugin and a small adjustment here, which should fix the unit problem of the byte metrics in the evolution chart. (Not yet sure, but maybe some tests need to be adjusted).
As discussed in the meeting ff you want feel free to have another look at the ecommerce sparklines. If you think it will take too long, feel free to postpone. Otherwise guess this PR should be almost finished.

File: plugins/Ecommerce/Controller.php
Comment (justinvelluppillai) [reply plugins/Ecommerce/Controller.php:118]: I'm late to the party here but I'd say we should review some of these issues and other refactoring as part of the 5.0.0 milestone coming soon. For now it would be good to get this PR merged as it solves other issues, though we'd need to have some idea of the time involved to best decide here. If it's 1 hour extra to refactor then it's probably worth doing, but if it doubles the time remaining or delays this PR from being merged then I'd agree with @bx80 that we should stick to the original time value of this task as much as possible.

**Comment by bx80** [comment]
@sgiehl Thanks for fixing the bandwidth metrics, that's looking really good  :smiley:
I don't think I'm going to have the time to rework the ecommerce sparklines for this release. 
I've tidied up the remaining tests. 

Not sure if you want to do a quick final review or shall I go ahead and merge?

**Comment by sgiehl** [comment]
@bx80 I have merge in the latest changed from `4.x-dev` to see if any tests are still failing. But looking through the changed UI tests it seems like the numbers for the sparklines aren't formatted anymore when comparing data. Haven't looked in detail yet. I will wait for the tests to finish, maybe they only need to be updated 🤔

**Comment by sgiehl** [comment]
@bx80 Seems you need to have another look at that one. The number formatting seems to fail when comparing data. I was able to reproduce that locally.

**Comment by bx80** [comment]
@sgiehl I'm not able to see where the number formatting is failing, the failed UI tests all seem to be caused by the action data tables not grouping and when I run this test locally they group correctly. I've also checked all the obvious places in the UI both in English and Deutsch (in case it's locale specific) but the sparkline tooltips number comparisons seem to be formatted correctly. I'm clearly missing something, where should I be looking? :eyes:

**Comment by sgiehl** [comment]
Are the average spent time and bounce rate metric formatted for you?

**Comment by bx80** [comment]
>Are the average spent time and bounce rate metric formatted for you?

Yes, they seem ok.
image
image
image

**Comment by sgiehl** [comment]
As mentioned before it seems only broken in comparison mode. So when comparing dates or segments.

**Comment by bx80** [comment]
@sgiehl Sorry, I was thinking comparing via the tooltips not via comparison mode. :man_facepalming: I see the issue now and I've commit a fix to also the format metrics when the sparklines are in comparison mode. 
image

File: plugins/Goals/Reports/Get.php
Comment (sgiehl) [inline plugins/Goals/Reports/Get.php]: There currently seems to be a problem with overall revenue. There is no evolution calculated for it. Guess the problem is, that the column is formatted using a filter above. And so it isn't numeric here anymore...

File: plugins/CoreVisualizations/Visualizations/Sparklines.php
Comment (sgiehl) [inline plugins/CoreVisualizations/Visualizations/Sparklines.php]: Guess using `Report::getProcessedMetricsForTable` would be more accurate.

File: plugins/CoreVisualizations/Visualizations/Sparklines.php
Comment (sgiehl) [inline plugins/CoreVisualizations/Visualizations/Sparklines.php]: Finding the correct metric for formatting can be done outside the nested foreach loops, no need to do that over and over again. Also this is actually not always accurate. Their can be multiple columns given, which might not necessarily have the same format. e.g. it could be a percent and money value. Currently both metrics would be formatted with the first matching one. Guess we need to format each separately...

File: plugins/CoreVisualizations/Visualizations/Sparklines.php
Comment (sgiehl) [inline plugins/CoreVisualizations/Visualizations/Sparklines.php]: Same as above

**Review by sgiehl — COMMENTED** [review]
Found some more small issues that don't work properly. Will try to push some fixes for it, so we can go on and merge this issue soon

**Comment by sgiehl** [comment]
@bx80 with the fixes I've pushed, I think this should be good to merge now. Feel free to have another quick look at the changes and merge it if they look fine for you.
Regarding the ecommerce sparkline refactoring. Found this issue https://github.com/matomo-org/matomo/issues/15349, which we should handle that in.

Btw. before merging this one, you need to merge all related submodule PRs and update the submodules in this PR back to their 4.x-dev branch again. Otherwise we might get submodule reference issues....

**Pull request merged**
